querier.go

```diff
@@ -262,37 +262,92 @@ func (q *blockQuerier) Close() error {
 }
 
 // PostingsForMatchers assembles a single postings iterator against the index reader
-// based on the given matchers. It returns a list of label names that must be manually
-// checked to not exist in series the postings list points to.
-// It returns EmptyPostings() if it can be determined beforehand that no results will be found.
+// based on the given matchers.
 func PostingsForMatchers(ix IndexReader, ms ...labels.Matcher) (index.Postings, error) {
-	var its []index.Postings
+	var its, notIts []index.Postings
+	// See which label must be non-empty.
+	labelMustBeSet := make(map[string]bool, len(ms))
+	for _, m := range ms {
+		if !m.Matches("") {
+			labelMustBeSet[m.Name()] = true
+		}
+	}
 
 	for _, m := range ms {
-		it, err := postingsForMatcher(ix, m)
+		matchesEmpty := m.Matches("")
+		if labelMustBeSet[m.Name()] || !matchesEmpty {
+			// If this matcher must be non-empty, we can be smarter.
+			nm, isNot := m.(*labels.NotMatcher)
+			if isNot && matchesEmpty { // l!="foo"
```

Comment:
are you sure about this one? `matchesEmpty` is when the label value is "" , no? the way I read it is `l!=""`

Reply:
l!="foo" matches an empty value.

Reply:
yes but than you have "isNot" which reverses the results.

Reply:
That means there's a Not present, l="foo" would not match empty.

Reply:
I did a quick test and that condition is entered for every non empty label value so I still don't see why the comment is  `// l!="foo"` and not `// l!=""`

https://github.com/prometheus/tsdb/compare/master...krasi-georgiev:TestNotMatcherPR?expand=1

Reply:
Hi, take `l != "foo"`, both `matchesEmpty` and `isNot` are true for it.

For `l != ""`, `matchesEmpty` is false no?

Reply:
IIUC Brian is suggesting that this comment means that the condition is entered when the label value is empty which is not the case.